Serialization / ipc_stream roundtrip errors with ComputeError: out-of-spec: ExpectedBuffer

### Checks

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import io
schema = {
        "s": pl.Utf8,
        "list": pl.List(pl.Utf8),
    }
empty = pl.DataFrame(
        {
            "s": [],
            "list": [],
        }
    ).cast(schema).lazy()

row1 = pl.DataFrame({"s" : ["A"] ,"list" : [[]] }).cast(schema).lazy()
row2 = pl.DataFrame({"s" : ["B"] ,"list" : [[]]}).cast(schema).lazy()
df = pl.concat([empty, row1, row2])

bytes = df.collect().serialize()
deserialized = pl.DataFrame.deserialize(io.BytesIO(bytes))
print(deserialized)
```


### Log output

```shell
if isinstance(source, StringIO):
            source = BytesIO(source.getvalue().encode())
        elif isinstance(source, (str, Path)):
            source = normalize_filepath(source)
    
        if format == "binary":
            deserializer = PyDataFrame.deserialize_binary
        elif format == "json":
            deserializer = PyDataFrame.deserialize_json
        else:
            msg = f"`format` must be one of {{'binary', 'json'}}, got {format!r}"
            raise ValueError(msg)
    
>       return cls._from_pydf(deserializer(source))
E       polars.exceptions.ComputeError: out-of-spec: ExpectedBuffer

cls        = <class 'polars.dataframe.frame.DataFrame'>
deserializer = <built-in method deserialize_binary of type object at 0x104711e70>
format     = 'binary'
source     = <_io.BytesIO object at 0x112698c70>
```

### Issue description

Deserialization of dataframe doesn't work sometimes in 1.22.0

### Expected behavior

Serialize -> deserialize should return original dataframe

### Installed versions

<details>
```
Polars:              1.22.0
Index type:          UInt32
Platform:            macOS-15.2-arm64-arm-64bit
Python:              3.10.12 (main, Oct 31 2023, 16:05:04) [Clang 15.0.0 (clang-1500.0.40.1)]
LTS CPU:             False

----Optional dependencies----
Azure CLI            <not installed>
adbc_driver_manager  <not installed>
altair               <not installed>
azure.identity       <not installed>
boto3                1.35.5
cloudpickle          3.0.0
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               <not installed>
gevent               <not installed>
google.auth          2.32.0
great_tables         <not installed>
matplotlib           3.9.2
numpy                2.1.0
openpyxl             <not installed>
pandas               <not installed>
pyarrow              <not installed>
pydantic             1.10.18
pyiceberg            <not installed>
sqlalchemy           <not installed>
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

--- a/crates/polars-arrow/src/compute/concatenate.rs
+++ b/crates/polars-arrow/src/compute/concatenate.rs
@@ -318,7 +318,6 @@ fn concatenate_view<V: ViewType + ?Sized, A: AsRef<dyn Array>>(
         Arc::new([Buffer::from(new_buffer)]) as Arc<[_]>
     };
 
-    dbg!(buffers.len());
     unsafe {
         BinaryViewArrayGeneric::new_unchecked(
             dtype,

--- a/crates/polars-arrow/src/compute/concatenate.rs
+++ b/crates/polars-arrow/src/compute/concatenate.rs
@@ -204,6 +204,9 @@ fn concatenate_view<V: ViewType + ?Sized, A: AsRef<dyn Array>>(
 ) -> BinaryViewArrayGeneric<V> {
     let dtype = arrays[0].as_ref().dtype().clone();
     let (total_len, null_count) = len_null_count(arrays);
+    if total_len == 0 {
+        return BinaryViewArrayGeneric::new_empty(dtype);
+    }
     let validity = concatenate_validities_with_len_null_count(arrays, total_len, null_count);
 
     let first_arr: &BinaryViewArrayGeneric<V> = arrays[0].as_ref().as_any().downcast_ref().unwrap();
@@ -315,6 +318,7 @@ fn concatenate_view<V: ViewType + ?Sized, A: AsRef<dyn Array>>(
         Arc::new([Buffer::from(new_buffer)]) as Arc<[_]>
     };
 
+    dbg!(buffers.len());
     unsafe {
         BinaryViewArrayGeneric::new_unchecked(
             dtype,

--- a/crates/polars-arrow/src/io/ipc/write/common.rs
+++ b/crates/polars-arrow/src/io/ipc/write/common.rs
@@ -260,7 +260,7 @@ fn set_variadic_buffer_counts(counts: &mut Vec<i64>, array: &dyn Array) {
         },
         ArrowDataType::LargeList(_) => {
             // Subslicing can change the variadic buffer count, so we have to
-            // slice here as well.
+            // slice here as well to stay synchronized.
             let array = array.as_any().downcast_ref::<LargeListArray>().unwrap();
             let offsets = array.offsets().buffer();
             let first = *offsets.first().unwrap();

--- a/crates/polars-arrow/src/io/ipc/write/common.rs
+++ b/crates/polars-arrow/src/io/ipc/write/common.rs
@@ -13,6 +13,7 @@ use crate::io::ipc::read::Dictionaries;
 use crate::legacy::prelude::LargeListArray;
 use crate::match_integer_type;
 use crate::record_batch::RecordBatchT;
+use crate::types::Index;
 
 /// Compression codec
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
@@ -258,8 +259,16 @@ fn set_variadic_buffer_counts(counts: &mut Vec<i64>, array: &dyn Array) {
             }
         },
         ArrowDataType::LargeList(_) => {
+            // Subslicing can change the variadic buffer count, so we have to
+            // slice here as well.
             let array = array.as_any().downcast_ref::<LargeListArray>().unwrap();
-            set_variadic_buffer_counts(counts, array.values().as_ref())
+            let offsets = array.offsets().buffer();
+            let first = *offsets.first().unwrap();
+            let last = *offsets.last().unwrap();
+            let subslice = array
+                .values()
+                .sliced(first.to_usize(), last.to_usize() - first.to_usize());
+            set_variadic_buffer_counts(counts, &*subslice)
         },
         ArrowDataType::FixedSizeList(_, _) => {
             let array = array.as_any().downcast_ref::<FixedSizeListArray>().unwrap();

--- a/py-polars/tests/unit/io/test_ipc.py
+++ b/py-polars/tests/unit/io/test_ipc.py
@@ -2,7 +2,7 @@ from __future__ import annotations
 
 import io
 from decimal import Decimal
-from typing import TYPE_CHECKING, Any
+from typing import TYPE_CHECKING, Any, no_type_check
 
 import pandas as pd
 import pytest
@@ -457,6 +457,7 @@ def test_enum_scan_21564() -> None:
     )
 
 
+@no_type_check
 def test_roundtrip_empty_str_list_21163() -> None:
     schema = {
         "s": pl.Utf8,

--- a/py-polars/tests/unit/io/test_ipc.py
+++ b/py-polars/tests/unit/io/test_ipc.py
@@ -455,3 +455,16 @@ def test_enum_scan_21564() -> None:
         pl.scan_ipc(f).collect().to_series(),
         s,
     )
+
+
+def test_roundtrip_empty_str_list_21163() -> None:
+    schema = {
+        "s": pl.Utf8,
+        "list": pl.List(pl.Utf8),
+    }
+    row1 = pl.DataFrame({"s": ["A"], "list": [[]]}).cast(schema)
+    row2 = pl.DataFrame({"s": ["B"], "list": [[]]}).cast(schema)
+    df = pl.concat([row1, row2])
+    bytes = df.serialize()
+    deserialized = pl.DataFrame.deserialize(io.BytesIO(bytes))
+    assert_frame_equal(df, deserialized)

--- a/py-polars/tests/unit/io/test_ipc.py
+++ b/py-polars/tests/unit/io/test_ipc.py
@@ -462,8 +462,8 @@ def test_roundtrip_empty_str_list_21163() -> None:
         "s": pl.Utf8,
         "list": pl.List(pl.Utf8),
     }
-    row1 = pl.DataFrame({"s": ["A"], "list": [[]]}).cast(schema)
-    row2 = pl.DataFrame({"s": ["B"], "list": [[]]}).cast(schema)
+    row1 = pl.DataFrame({"s": ["A"], "list": [[]]}, schema=schema)
+    row2 = pl.DataFrame({"s": ["B"], "list": [[]]}, schema=schema)
     df = pl.concat([row1, row2])
     bytes = df.serialize()
     deserialized = pl.DataFrame.deserialize(io.BytesIO(bytes))

